src/document-fetcher.js

```diff
@@ -0,0 +1,136 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {Services} from './services';
+import {assertSuccess, getViewerInterceptResponse, setupAMPCors, setupInit, setupInput, verifyAmpCORSHeaders} from './utils/xhr-utils';
+import {dict} from './utils/object';
+import {user} from './log';
+
+/**
+ *
+ *
+ * @param {!Window} win
+ * @param {string} input
+ * @param {?./utils/xhr-utils.FetchInitDef=} opt_init
+ * @return {!Promise<!Document>}
+ * @ignore
+ */
+export function fetchDocument(win, input, opt_init) {
+  let init = setupInit(opt_init, 'text/html');
+  init = setupAMPCors(win, input, init);
+  input = setupInput(win, input, init);
+  const ampdocService = Services.ampdocServiceFor(win);
+  const ampdocSingle_ =
+  ampdocService.isSingleDoc() ? ampdocService.getAmpDoc() : null;
+  init.responseType = 'document';
+  return getViewerInterceptResponse(win, ampdocSingle_, input, init)
+      .then(interceptorResponse => {
+        if (interceptorResponse) {
+          return interceptorResponse.text().then(body =>
+            new DOMParser().parseFromString(body, 'text/html')
+          );
+        }
+        return xhrRequest_(input, init).then(({xhr, response}) => {
+          verifyAmpCORSHeaders(win, response, init);
+          return xhr.responseXML;
+        });
+      });
+}
+
+/**
+ *
+ *
+ * @param {string} input
+ * @param {!./utils/xhr-utils.FetchInitDef} init
+ * @private
+ */
+function xhrRequest_(input, init) {
+  return new Promise((resolve, reject) => {
+    const xhr = new XMLHttpRequest();
+    xhr.open(init.method || 'GET', input, true);
+    xhr.withCredentials = (init.credentials == 'include');
+    xhr.responseType = 'document';
+    // Incoming headers are in fetch format,
+    // so we need to convert them into xhr.
+    if (init.headers) {
+      for (const header in init.headers) {
+        xhr.setRequestHeader(header, init.headers[header]);
+      }
+    }
+    xhr.onreadystatechange = () => {
+      if (xhr.readyState < /* STATUS_RECEIVED */ 2) {
+        return;
+      }
+      if (xhr.status < 100 || xhr.status > 599) {
+        xhr.onreadystatechange = null;
+        reject(user().createExpectedError(
+            `Unknown HTTP status ${xhr.status}`));
+        return;
+      }
+      // TODO(dvoytenko): This is currently simplified: we will wait for the
+      // whole document loading to complete. This is fine for the use cases
+      // we have now, but may need to be reimplemented later.
+      if (xhr.readyState == /* COMPLETE */ 4) {
+        const options = {
+          status: xhr.status,
+          statusText: xhr.statusText,
+          headers: parseHeaders(xhr.getAllResponseHeaders()),
+        };
+        const body = 'response' in xhr
+          ? xhr.response : xhr.responseText;
+        const response = new Response(/** @type {string} */ (body || ''), /** @type {!ResponseInit} */ (options));
```

Comment:
Actually this is done because at the end of all the change(including fetch polyfill), everything should talk in terms of native `Response`, `Request`. So we'd have to  construct Response to talk to `assertSuccess`, `verifyAmpCORSHeaders` etc.

Reply:
I disagree. Sure, we can make things that need a response to work (like, getting the response's body) use a `Response`. But `verifyAmpCORSHeaders` only needs a `Headers` (and only really needs `{ get() {} }`. Simplifying the type requirements makes this code easier to reason about, because we're never going to use the response's body (`xhr.reponseText`) here, we aren't going to use `status` or `statusText`.

I really think we can simplify code by resolving with just the necessities here:

```js
const makeshift = {
  xml: xhr.responseXML,
  // We can make this a wrapper class if we want to provide the full API
  headers: {
    get(header) {
      return xhr.getResponseHeader(header);
    },
  },
};
resolve(makeshift);

// back to https://github.com/ampproject/amphtml/pull/17259/files#diff-71317e0e89806231df17433535139205R45
// callback parameter is already a struct, not a Response
return xhrRequest(input, init).then(({xml, headers}) => {
  verifyAmpCORSHeaders(win, headers, init);
  return xml;
});

Reply:
Woops, I actually noticed something.
Till this PR nobody has polyfilled Response for IE11 and other browsers so I am changing this to FetchResponse anyways.

But I will come back to this after doing the polyfill PR. 
Note for then: 
`verifyAmpCORSHeaders` requires the headers part
`assertSuccess` requires the `.ok`. So either we make both of them not take `response` arguments at all or we construct the `Response` once and use it everywhere.